src/pcp/buddyinfo/pcp-buddyinfo.1
'\" t
.\"
.\" Man page for pcp-buddyinfo
.\" Copyright (c) 2023 Oracle and/or its affiliates.
.\" DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.TH PCP-BUDDYINFO 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pcp-buddyinfo\f3 \- report Linux kernel buddy algorithm statistics
.SH SYNOPSIS
\f3pcp\f1
[\f2pcp\ options\f1]
\f3buddyinfo\f1
[\f3\-uVz?\f1]
[\f3\-s\f1 \f2samples\f1]
[\f3\-a\f1 \f2archive\f1]
[\f3\-Z\f1 \f2timezone\f1]
.SH DESCRIPTION
The
.B pcp-buddyinfo
command is used for viewing different stats related to buddyinfo.
It helps users analyze the dynamic behaviour of the buddy algorithm
used in the Linux kernel virtual memory subsystem.
The information includes the total number of zones that are currently
active, counts of different order pages, and so on.
By default,
.B pcp-buddyinfo
reports live data for the local host.
.PP
The statistics shown are as follows:

.TS
lfB  lfB
l    lx.
HEADER          DESCRIPTION
_               _
Normal          zones available
Nodes           available nodes
Order0          available pages of order 0
Order1          available pages of order 1
Order2          available pages of order 2
Order3          available pages of order 3
Order4          available pages of order 4
Order5          available pages of order 5
Order6          available pages of order 6
Order7          available pages of order 7
Order8          available pages of order 8
Order9          available pages of order 9
Order10         available pages of order 10
.TE


Each column represents the number of pages of a certain order (a certain size) that are available at any given time. For example, for zone DMA (direct memory access), there are 90 of 2^(0*PAGE_SIZE) chunks of memory. Similarly, there are 6 of 2^(1*PAGE_SIZE) chunks, and 2 of 2^(2*PAGE_SIZE) chunks of memory available.
.PP
The DMA row references the first 16 MB on a system, the HighMem row references all memory greater than 4 GB on a system, and the Normal row references all memory in between.
.SH OPTIONS
.TP
\fB\-a\fR \fIfile\fR, \fB\-\-archive\fR=\fIfile\fR
Fetch buddyinfo metrics from the specified PCP archive.
.TP
\fB\-s\fR \fIcount\fR, \fB\-\-samples\fR=\fIcount\fR
Limit reporting to \fIcount\fR samples of buddyinfo statistics.
.TP
\fB\-u\fR, \fB\-\-no-interpol\fR
disable the default interpolation mode when replaying an archive.
With the
.B \-u
option, interpolated reporting is disabled and each sample is
reported according to native sampling intervals in the archive.
In this mode the
.B \-t
option is ignored.
Additionally, non-interpolated replay makes sense only when
replaying an archive (using the
.B \-a
option described above), so the
.B \-a
option is mandatory when
.B \-u
is specified.
.TP
\fB\-z\fR, \fB\-\-hostzone\fR
Use the local timezone of the host that is the source of the
performance metrics, as identified by either the
.B \-h
or the
.B \-a
options.
The default is to use the timezone of the local host.
.TP
\fB\-Z\fR \fItimezone\fR, \fB\-\-timezone\fR=\fItimezone\fR
Use
.I timezone
for the date and time.
.I Timezone
is in the format of the environment variable
.B TZ
as described in
.BR environ (7).
.TP
\fB\-V\fR, \fB\-\-version\fR
Display version number and exit.
.TP
\fB\-?\fR, \fB\-\-help\fR
Display usage message and exit.
.SH NOTES
.B pcp-buddyinfo
reports information extracted from the \fI/proc/buddyinfo\fP Linux
kernel procfs file.
.PP
If the default interpolation mode is disabled, it is possible that
some metrics (recorded at different intervals) will be reported as
having missing values for some samples even if they were recorded.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize the file and directory names used by PCP. On each installation, the file \fI/etc/pcp.conf\fP contains the local values for these variables. The \fB$PCP_CONF\fP variable may be used to specify an alternative configuration file, as described in \fBpcp.conf\fP(5).
.PP
For environment variables affecting PCP tools, see \fBpmGetOptions\fP(3).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pcp (1),
.BR pmGetOptions (3),
.BR pcp.conf (5)
and
.BR environ (7).

.\" control lines for scripts/man-spell
.\" +ok+ PAGE_SIZE buddyinfo {from pcp-buddyinfo} HighMem


CHANGELOG
pcp-7.0.0 (1 July 2025)
    - Work-in-progress.

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out
	build: update denki.conf packaging location for consistency
    - Documentation:
	man pages: tidy up references to $PCP_DEBUG
	man pages: __pmHash* added to man3i
	man pages: pmnewcontext(3) unix: and local: option details
	docs: man page updates transitioning away from Redis
	docs: correct pmlogger(1) man page reference to section 5 page
	docs: update the books to refer to key servers more generically
	pmrep et al man pages: describe derived metric troubleshooting

pcp-6.3.0 (30 July 2024)
    - Client tools and utilities:
	pcp-htop: add AMD GPU metrics support
	pcp-htop: fix Available Meter column list in Setup/F2
	pcp-htop: fix nested derived metric definitions for devices
	pcp-htop: fix label/color botch in the MySQL keys Meter config
	pcp-xsos: fast system overview for Red Hat customer support
	pcp-atop: fix units conversion for ipc.shm metric values
	pcp2openmetrics: stricter adherence to the specification
	pmview: developer preview only, incl. mpvis, dkvis, osvis
	pmstat: include guest time in extra CPU stats
	pmrep conf: add support for pmstat -x alike metricset
	pmrep conf: fix collectl cpu metrics to match current versions
	pmrep conf: add guest cpu time to vmstat output where missing
	pmrep conf: add metricset to mimic pmstat(1)
	pmrep conf: fix cpu metrics in sar/vmstat metricset formulas
	pmconfig: add -L y2038_safe to know if 8-byte time_t used
    - PMDA additions, enhancements and bug fixes:
	pmdaamdgpu: metrics from libdrm and libdrm-amdgpu libraries

src/pcp/buddyinfo/pcp-buddyinfo.py
#!/usr/bin/env pmpython
#
# Copyright (c) 2023 Oracle and/or its affiliates.
# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# pylint: disable=bad-whitespace,too-many-arguments,too-many-lines, bad-continuation
# pylint: disable=redefined-outer-name,unnecessary-lambda
#
import signal
import sys
import time
from pcp import pmapi, pmcc
from cpmapi import PM_CONTEXT_ARCHIVE, PM_MODE_FORW

SYS_METRICS= ["kernel.uname.sysname","kernel.uname.release",
               "kernel.uname.nodename","kernel.uname.machine","hinv.ncpu"]
BUDDYSTAT_METRICS = ["mem.buddyinfo.pages","mem.buddyinfo.total"]

ALL_METRICS = BUDDYSTAT_METRICS + SYS_METRICS

def adjust_length(name,size):
    return name.ljust(size)

class ReportingMetricRepository:

    def __init__(self,group):
        self.group=group
        self.current_cached_values = {}

    def __fetch_current_value(self,metric):
        val=dict(map(lambda x: (x[1], x[2]), self.group[metric].netValues))
        return dict(val)

    def current_value(self,metric):
        if metric not in self.group:
            return None
        if self.current_cached_values.get(metric) is None:
            first_value=self.__fetch_current_value(metric)
            self.current_cached_values[metric]=first_value
        return self.current_cached_values[metric]

class BuddyStatUtil:
    def __init__(self,metrics_repository):
        self.__metric_repository=metrics_repository
        self.report=ReportingMetricRepository(self.__metric_repository)

    def buddy_pages(self):
        return self.report.current_value('mem.buddyinfo.pages')

    def buddy_total_pages(self):
        return self.report.current_value('mem.buddyinfo.total')

    def names(self):
        data = self.report.current_value('mem.buddyinfo.pages')
        return data.keys()

class BuddyinfoReport(pmcc.MetricGroupPrinter):
    def __init__(self,opts,group):
        self.opts=opts
        self.group=group
        self.context=opts.context
        self.samples=opts.samples
        self.header = "unknown"

    def __get_ncpu(self, group):
        return group['hinv.ncpu'].netValues[0][2]

    def __print_machine_info(self, context):
        if self.header == "unknown":
            try:
                timestamp = self.group.pmLocaltime(context.timestamp.tv_sec)
                # Please check strftime(3) for different formatting options.
                # Also check TZ and LC_TIME environment variables for more
                    # information on how to override the default formatting of
                    # the date display in the header
                time_string = time.strftime("%x", timestamp.struct_time())
                header_string = ''
                header_string += context['kernel.uname.sysname'].netValues[0][2] + '  '
                header_string += context['kernel.uname.release'].netValues[0][2] + '  '
                header_string += '(' + context['kernel.uname.nodename'].netValues[0][2] + ')  '
                header_string += time_string + '  '
                header_string += context['kernel.uname.machine'].netValues[0][2] + '  '
                self.header = header_string
            except  IndexError:
                pass
        try:
            print("%s  (%s CPU)" % (self.header, self.__get_ncpu(context)))
        except  IndexError:
            pass

    def __print_header(self,header_indentation,value_indentation):
        value_indentation+=" "*2
        print("TimeStamp"+ " "+header_indentation + "Normal" + header_indentation+value_indentation+" " +
              "Nodes" + header_indentation + "Order0" + value_indentation +
              "Order1" + value_indentation +"Order2" + value_indentation + "Order3" + value_indentation +
              "Order4" + value_indentation +"Order5" + value_indentation + "Order6" + value_indentation +
              "Order7" + value_indentation +"Order8" + value_indentation + "Order9" +
              value_indentation + "Order10")

    def __print_values(self,timestamp,header_indentation,\
                     value_indentation,buddystatus):
        names=buddystatus.names()
        pages=buddystatus.buddy_pages()
        order_set = set()
        nodes_set = set()
        no_of_nodes_set = set()
        for name in names:
            part=name.split('::')
            if len(part)==3:
                nodes_set.add(part[0])
                order_set.add(part[1])
                no_of_nodes_set.add(part[2])
        def __extract_numeric_part(element):
            return int(element[5:])
        for Normal in sorted(nodes_set):
            for node in no_of_nodes_set:
                value=""
                for order in sorted(order_set,key=__extract_numeric_part):
                    nodename = adjust_length(Normal,9) if len(Normal) < 9 else Normal
                    key = "%s::%s::%s" % (Normal, order, node)
                    data = str(pages.get(key,0))
                    value +=  adjust_length(data,8) if len(data) < 8 else data
                    value += value_indentation

                print("%s %s %s %s %s %s %s "%(timestamp,header_indentation,nodename,header_indentation,node,
                                                  header_indentation,value))

    def print_report(self,group,timestamp,header_indentation,value_indentation):

        def __print_buddy_status():
            buddystatus = BuddyStatUtil(group)
            if buddystatus.names():
                try:
                    self.__print_machine_info(group)
                    self.__print_header(header_indentation, value_indentation)
                    self.__print_values(timestamp, header_indentation, value_indentation, buddystatus)
                except IndexError:
                    print("Incorrect machine info due to some missing metrics")
                return
            else:
                return

        if self.context != PM_CONTEXT_ARCHIVE and self.samples is None:
            __print_buddy_status()
            sys.exit(0)
        elif self.context == PM_CONTEXT_ARCHIVE and self.samples is None:
            __print_buddy_status()
        elif self.samples >=1:
            __print_buddy_status()
            self.samples-=1

    def report(self, manager):
        group = manager["allinfo"]
        self.samples=self.opts.pmGetOptionSamples()
        t_s = group.contextCache.pmLocaltime(int(group.timestamp))
        timestamp = time.strftime(BuddyinfoOptions.timefmt, t_s.struct_time())
        header_indentation = "        " if len(timestamp) < 9 else (len(timestamp) - 7) * " "
        value_indentation = ((len(header_indentation) + 2) - len(timestamp)) * " "
        self.print_report(group,timestamp,header_indentation,value_indentation)

class BuddyinfoOptions(pmapi.pmOptions):
    timefmt = "%H:%M:%S"
    uflag = False

    def extraOptions(self,opt,optarg,index):
        if opt == 'u':
            BuddyinfoOptions.uflag = True

    def __init__(self):
        pmapi.pmOptions.__init__(self, "a:s:Z:uzV?")
        self.pmSetLongOptionHeader("General options")
        self.pmSetOptionCallback(self.extraOptions)
        self.pmSetLongOptionArchive()
        self.pmSetLongOption("no-interpol", 0, "u", "", "disable interpolation mode with archives")
        self.pmSetLongOptionHostZone()
        self.pmSetLongOptionTimeZone()
        self.pmSetLongOptionHelp()
        self.pmSetLongOptionSamples()
        self.pmSetLongOptionVersion()
        self.samples=None
        self.context=None
    def checkOptions(self, manager):
        if BuddyinfoOptions.uflag:
            if manager._options.pmGetOptionInterval():  # pylint: disable=protected-access
                print("Error: -t incompatible with -u")
                return False
            if manager.type != PM_CONTEXT_ARCHIVE:
                print("Error: -u can only be specified with -a archive")
                return False
        return True

if __name__ == '__main__':
    try:
        opts = BuddyinfoOptions()
        mngr = pmcc.MetricGroupManager.builder(opts,sys.argv)
        opts.context=mngr.type

        if not opts.checkOptions(mngr):
            raise pmapi.pmUsageErr
        if BuddyinfoOptions.uflag:
            # -u turns off interpolation
            mngr.pmSetModeHighRes(PM_MODE_FORW, mngr._options.pmGetOptionOrigin(), None) # pylint: disable=protected-access

        missing = mngr.checkMissingMetrics(ALL_METRICS)
        if missing is not None:
            sys.stderr.write('Error: not all required metrics are available\nMissing: %s\n' % (missing))
            sys.exit(1)
        mngr["buddyinfo"] = BUDDYSTAT_METRICS
        mngr["allinfo"]=ALL_METRICS
        mngr.printer = BuddyinfoReport(opts,mngr)
        sts = mngr.run()
        sys.exit(sts)
    except pmapi.pmErr as error:
        sys.stderr.write('%s\n' % (error.message()))
    except pmapi.pmUsageErr as usage:
        usage.message()
        sys.exit(1)
    except IOError:
        signal.signal(signal.SIGPIPE, signal.SIG_DFL)
    except KeyboardInterrupt:
        pass
